### PR TITLE
Initial implementations for SSH key & host management views

### LINKED ISSUE #2: Figure out how to make specific views require login from within Flask
We don't want to have to protect locations in the Apache config. Instead we should have a decorator for views that redirects to `/login?target=` if the user is not logged in, and back to the same view when they do.

Can use `redirect(url_for('login', target='/'))` possibly; not sure if flask-sso will do the redirect for us automatically. Doesn't look like it does, so our login handler will need to do so. Need to figure out how to read the target parameter.

### LINKED ISSUE #5: Design a database structure for the front-end
Needs to have tables at least for:
* Users: UID (pk), UPI (maybe - we can get that from the session), ...
* SSH keys: id (pk), user (fk), public_key, label
* Virtual hosts: id (pk), user (fk), name (unique per user), DNS name (unique), description (may make name redundant?), ...
    * username & password, or SSH key (fk) - for logging in to the machine
    * Terraform state file (JSON) - canonical source of config details, can be refreshed by backend API call

Since hosts can be set up in different ways, we may need some other tables for their config, e.g. a table for Dockerfiles. User needs to fill in for initial creation:
* VM size, base OS, etc from available options (initially fixed)
* Script (sequence of shell commands) to run on the VM (as root?) to set it up
* Port to connect to for testing the service is up & working, and to include in the link to their new website

Maybe another table for user-GitHub repo links? Possibly using python-social-auth here like Dashboard does? Not needed for initial demo, but will be to support private repos #6.

Initial idea was that front-end will query back-end web service to determine machine status when showing the dashboard - this doesn't need to be in the DB. It'll pass a list of unique machine IDs (DNS name?) and get back statuses, including 'this machine has been deleted via the Azure portal' telling it to drop the machine from the DB (or possibly not, in case user wants to recreate it easily?). This may now be done better by refreshing the Terraform state file.

### PR DESCRIPTION
**Pull request by jonc125**
At present the host management just updates our DB, rather than doing anything on Azure. That will be for a separate PR - here the focus is on UI.

Also fixes #2, fixes #5.

### CODE REVIEW

File: requirements/base.txt [hunk 1]
```diff
@@ -9,8 +9,9 @@ blinker==1.4              # via flask-sso
 click==6.7                # via flask
```
Comment (raquelalegre) [inline requirements/base.txt:9]: I've seen people do requirements like this but I have never done it myself. Why do we need a separate base and local requirements? I can't see a big difference between these too. There's pip-tools and first which are only in the local one.

File: src/cloudlabs/forms/add_host.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        validators=[v.Optional()])
+    dns_name = StringField(
+        'URL',
+        description='The canonical DNS name for this host, which must be globally unique',
+        validators=[v.Length(min=1, max=Host.dns_name.type.length)])
+
+    admin_username = StringField(
+        'Admin username',
+        description='Username of the admin user to be created on the new host',
+        validators=[v.Length(min=1, max=Host.admin_username.type.length)])
```
Comment (raquelalegre) [inline src/cloudlabs/forms/add_host.py:27]: Is it possible to add validation for no spaces or symbols?

File: src/cloudlabs/forms/add_host.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        'Admin password',
+        description='Password to use for the admin user on this host',
+        validators=[v.Optional(), v.Length(min=12, max=Host.admin_password.type.length)])
+
+    git_repo = StringField(
+        'Git repository',
+        description='Location of the git repository to clone on the new host.'
+        ' It is assumed to contain a Dockerfile that will be built,'
+        ' although this can be customised.',
+        validators=[v.Length(min=1, max=Host.git_repo.type.length)])
```
Comment (raquelalegre) [inline src/cloudlabs/forms/add_host.py:47]: is it possible to add a regex for validation?

File: src/cloudlabs/forms/add_host.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        ' This is used to generate our default installation script,'
+        ' which can be customised if needed.',
+        default=80,
+        validators=[v.NumberRange(min=1, max=65535)])
+
+    def validate_label(form, field):
+        """Check a user doesn't duplicate labels."""
+        label = field.data.strip()
+        if Host.query.filter_by(user_id=g.user.id, label=label).count() > 0:
+            raise v.ValidationError('You must choose a unique label for each host')
```
Comment (raquelalegre) [inline src/cloudlabs/forms/add_host.py:60]: Also I think no spaces or symbols can be used.

File: src/cloudlabs/forms/add_key.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+class AddKeyForm(FlaskForm):
+    """Form for adding a new SSH public key."""
+    label = StringField(
+        'Label',
+        description='A label to allow you to identify this key',
+        validators=[v.Length(min=2, max=SshKey.label.type.length)])
+    public_key = TextAreaField(
+        'Public key',
+        description='The public key itself, e.g. copied from ~/.ssh/id_rsa.pub',
+        validators=[v.DataRequired()])
```
Comment (raquelalegre) [inline src/cloudlabs/forms/add_key.py:18]: Regex?

File: src/cloudlabs/templates/includes/_dashboard.html [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        {% endif %}
+        {% if host.status == 'Stopped' %}
+          <a href="{{ url_for('host.control', id=host.id, action='start') }}"><img src="{{ url_for('static', filename='img/start.svg') }}" alt="Start virtual machine" class="action" /></a>
+        {% endif %}
+        {# TODO: Add a 'Deploying' status with action to view log #}
+        <a href="{{ url_for('host.edit', id=host.id) }}"><img src="{{ url_for('static', filename='img/configure.svg') }}" alt="Configure virtual machine') }}" class="action" /></a>
+        <a href="{{ url_for('host.download', id=host.id) }}"><img src="{{ url_for('static', filename='img/download.svg') }}" alt="Download virtual machine image" class="action" /></a>
+        <img src="{{ url_for('static', filename='img/delete.svg') }}" alt="Delete virtual machine" class="action" onclick="confirm_delete('{{ url_for('host.delete', id=host.id) }}')" />
+        <a href="{{ url_for('host.info', id=host.id) }}"><img src="{{ url_for('static', filename='img/info.svg') }}" alt="Show virtual machine info" class="action" /></a>
```
Comment (raquelalegre) [inline src/cloudlabs/templates/includes/_dashboard.html:38]: Is the info button supposed to show the description of the item? Or is that at a later stage?

**Review by raquelalegre — APPROVED** [review]
All good! Tried it locally following README instructions and all went fine: I could create and delete hosts and add SSH keys on the Test User profile.

File: requirements/base.txt (hunk 1 above)
Comment (jonc125) [reply requirements/base.txt:9]: It's mostly future-proofing, when we start adding extra requirements for testing etc. they can go in a separate requirements file from that used by the server. I'm also using pip-tools to generate pinned requirements files from input specifications, so I don't need to track dependencies manually. So the .in files are the master, and the .txt files get auto-generated when you run make.

File: src/cloudlabs/forms/add_host.py (hunk 2 above)
Comment (jonc125) [reply src/cloudlabs/forms/add_host.py:27]: Probably; should do this as part of #1.

File: src/cloudlabs/templates/includes/_dashboard.html (hunk 6 above)
Comment (jonc125) [reply src/cloudlabs/templates/includes/_dashboard.html:38]: That's what the wireframes suggest. At the moment it's a fairly sparse description!

File: requirements/base.txt (hunk 1 above)
Comment (raquelalegre) [reply requirements/base.txt:9]: I should be that as well! Thanks :)

**Pull request merged**